Go should verify gziped messages are correctly terminated

Gzip has a footer containing an adler32 (think CRC32). grpc-dotnet recently introduced a bug where this was missing (https://github.com/grpc/grpc/issues/20884). However, the Go client did not fail. The gzip decoder should have failed with ErrChecksum. It's unclear why this wasn't reported to the client, but seems like a bug. (I poked around a bit and rpc_util.go's decompress() looked fine. The gzip decompressor itself didn't seem to have much to it.)

@ejona86 are you sure this doesn't already fail as expected?  I tried adding a test that uses a custom compressor which slightly modifies the final 4 bytes, and I see the following result:

```
--- FAIL: Test/GzipWithoutFooter (0.00s)
        end2end_test.go:7569: ss.client.UnaryCall(_) = _, rpc error: code = Internal desc = grpc: failed to decompress the received message gzip: invalid checksum; want _, nil
```

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -7537,7 +7537,7 @@ func (badGzipCompressor) Do(w io.Writer, p []byte) error {
 	}
 	err := gzw.Close()
 	bs := buf.Bytes()
-	bs[len(bs)-4] -= 1 // modify checksum (big endian) at end by 1 byte
+	bs[len(bs)-4]-- // modify checksum (big endian) at end by 1 byte
 	buf = bytes.NewBuffer(bs)
 	buf.WriteTo(w)
 	return err

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -7537,8 +7537,8 @@ func (badGzipCompressor) Do(w io.Writer, p []byte) error {
 	}
 	err := gzw.Close()
 	bs := buf.Bytes()
-	if len(bs) >= 4 {
-		bs[len(bs)-4]-- // modify checksum (big endian) at end by 1 byte
+	if len(bs) >= 6 {
+		bs[len(bs)-6] ^= 1 // modify checksum at end by 1 byte
 	}
 	w.Write(bs)
 	return err

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -7537,9 +7537,10 @@ func (badGzipCompressor) Do(w io.Writer, p []byte) error {
 	}
 	err := gzw.Close()
 	bs := buf.Bytes()
-	bs[len(bs)-4]-- // modify checksum (big endian) at end by 1 byte
-	buf = bytes.NewBuffer(bs)
-	buf.WriteTo(w)
+	if len(bs) >= 4 {
+		bs[len(bs)-4]-- // modify checksum (big endian) at end by 1 byte
+	}
+	w.Write(bs)
 	return err
 }
 
@@ -7547,7 +7548,7 @@ func (badGzipCompressor) Type() string {
 	return "gzip"
 }
 
-func (s) TestGzipWithoutFooter(t *testing.T) {
+func (s) TestGzipBadChecksum(t *testing.T) {
 	ss := &stubServer{
 		unaryCall: func(ctx context.Context, _ *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
 			return &testpb.SimpleResponse{}, nil
@@ -7563,9 +7564,11 @@ func (s) TestGzipWithoutFooter(t *testing.T) {
 
 	p, err := newPayload(testpb.PayloadType_COMPRESSABLE, int32(1024))
 	if err != nil {
-		panic(err)
+		t.Fatalf("Unexpected error from newPayload: %v", err)
 	}
-	if _, err := ss.client.UnaryCall(ctx, &testpb.SimpleRequest{Payload: p}); err == nil || status.Code(err) != codes.Internal {
-		t.Errorf("ss.client.UnaryCall(_) = _, %v; want _, Code()=codes.Internal", err)
+	if _, err := ss.client.UnaryCall(ctx, &testpb.SimpleRequest{Payload: p}); err == nil ||
+		status.Code(err) != codes.Internal ||
+		!strings.Contains(status.Convert(err).Message(), gzip.ErrChecksum.Error()) {
+		t.Errorf("ss.client.UnaryCall(_) = _, %v\n\twant: _, status(codes.Internal, contains %q)", err, gzip.ErrChecksum)
 	}
 }

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -24,6 +24,7 @@ package test
 import (
 	"bufio"
 	"bytes"
+	"compress/gzip"
 	"context"
 	"crypto/tls"
 	"errors"
@@ -7524,4 +7525,47 @@ func (s) TestClientCancellationPropagatesUnary(t *testing.T) {
 		t.Fatalf("server failed to close done chan due to cancellation propagation")
 	}
 	wg.Wait()
+}
+
+type badGzipCompressor struct{}
+
+func (badGzipCompressor) Do(w io.Writer, p []byte) error {
+	buf := &bytes.Buffer{}
+	gzw := gzip.NewWriter(buf)
+	if _, err := gzw.Write(p); err != nil {
+		return err
+	}
+	err := gzw.Close()
+	bs := buf.Bytes()
+	bs[len(bs)-4] -= 1 // modify checksum (big endian) at end by 1 byte
+	buf = bytes.NewBuffer(bs)
+	buf.WriteTo(w)
+	return err
+}
+
+func (badGzipCompressor) Type() string {
+	return "gzip"
+}
+
+func (s) TestGzipWithoutFooter(t *testing.T) {
+	ss := &stubServer{
+		unaryCall: func(ctx context.Context, _ *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
+			return &testpb.SimpleResponse{}, nil
+		},
+	}
+	if err := ss.Start(nil, grpc.WithCompressor(badGzipCompressor{})); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	p, err := newPayload(testpb.PayloadType_COMPRESSABLE, int32(1024))
+	if err != nil {
+		panic(err)
+	}
+	if _, err := ss.client.UnaryCall(ctx, &testpb.SimpleRequest{Payload: p}); err == nil || status.Code(err) != codes.Internal {
+		t.Errorf("ss.client.UnaryCall(_) = _, %v; want _, Code()=codes.Internal", err)
+	}
 }

